### PR TITLE
always expect exists filter

### PR DESCRIPTION
**Pull request by ppcad**
See issue #435

### CODE REVIEW

File: logprep/framework/rule_tree/rule_parser.py [hunk 1]
```diff
@@ -126,15 +130,36 @@ def _add_exists_filter(parsed_rules: list):
         for parsed_rule in parsed_rules:
             temp_parsed_rule = parsed_rule.copy()
             skipped_counter = 0
-
-            for segment_index, segment in enumerate(temp_parsed_rule):
```
Comment (dtrai2) [inline logprep/framework/rule_tree/rule_parser.py:130]: Could you maybe flatten the if-statements in this loop? That way it is a bit easier to understand. A suggestion cold look like this:

```python
  for segment_idx, segment in enumerate(temp_parsed_rule):
      if isinstance(segment, (Exists, Always)):
          skipped_counter += 1
          continue
      exists_filter = RuleParser._get_exists_filter(segment)
      if exists_filter is None:
          skipped_counter += 1
          continue
      if exists_filter in filter_expression:
          skipped_counter += 1
          continue
      filter_expression.insert(segment_idx * 2 - skipped_counter, exists_filter)
```

File: logprep/framework/rule_tree/rule_parser.py [hunk 2]
```diff
@@ -126,15 +130,36 @@ def _add_exists_filter(parsed_rules: list):
...
                     continue
 
-                exists_filter = Exists(segment.key)
-                if exists_filter in parsed_rule:
+                exists_filter = RuleParser._get_exists_filter(segment)
+                if exists_filter is None or exists_filter in parsed_rule:
                     skipped_counter += 1
-                    continue
+                else:
+                    parsed_rule.insert(segment_idx * 2 - skipped_counter, exists_filter)
```
Comment (dtrai2) [inline logprep/framework/rule_tree/rule_parser.py:142]: Can you explain where the 2 is coming from? I was trying to understand this and my first idea was that `segment_idx - 1` should work as well, but it doesn't. Maybe you could give this magic number a meaningful name?

File: logprep/framework/rule_tree/rule_parser.py [hunk 3]
```diff
@@ -126,15 +130,36 @@ def _add_exists_filter(parsed_rules: list):
...
+            Returns an Exists FilterExpression from a KeyValueBased FilterExpression or None.
+
+        """
+        if isinstance(segment, Not):
+            not_child = segment.children[0]
+            if isinstance(not_child, KeyValueBasedFilterExpression):
+                return Exists(not_child.key)
+        elif isinstance(segment, KeyValueBasedFilterExpression):
+            return Exists(segment.key)
+        return None
```
Comment (dtrai2) [inline logprep/framework/rule_tree/rule_parser.py:165]: How about we use this implementation? It's a bit easier to read and understand by preventing nested if-statements.
```suggestion
        if isinstance(segment, Not):
            segment = segment.children[0]
        if isinstance(segment, KeyValueBasedFilterExpression):
            return Exists(segment.key)
        return None
```

**Review by dtrai2 — CHANGES_REQUESTED** [review]

File: logprep/framework/rule_tree/rule_parser.py (hunk 1 above)
Comment (ppcad) [reply logprep/framework/rule_tree/rule_parser.py:130]: I've changed it.

File: logprep/framework/rule_tree/rule_parser.py (hunk 3 above)
Comment (ppcad) [reply logprep/framework/rule_tree/rule_parser.py]: Thanks for the suggestion. I've changed it.

File: logprep/framework/rule_tree/rule_parser.py (hunk 2 above)
Comment (ppcad) [reply logprep/framework/rule_tree/rule_parser.py]: It was a way to account for the new elements added to the list.

`segment_idx * 2 - skipped_counter` 
<=> `segment_idx - (segment_idx  - skipped_counter)`
<=> `segment_idx - added_exists_filter_count`

It was indeed confusing, so I changed it.

**Review by dtrai2 — APPROVED** [review]
Ah thanks for the changes, now the inserting is a bit easier to understand :+1:

**Pull request merged**